packages/worker/src/logger.ts

```diff
@@ -87,3 +87,68 @@ export class DefaultLogger implements Logger {
     this.log('ERROR', message, meta);
   }
 }
+
+export function withMetadata(logger: Logger, meta: LogMetadata | (() => LogMetadata)): Logger {
+  return new LoggerWithMetadata(logger, meta);
+}
+
+class LoggerWithMetadata implements Logger {
+  private parentLogger: Logger;
+  private metaChain: (LogMetadata | (() => LogMetadata))[];
+
+  constructor(parent: Logger, meta: LogMetadata | (() => LogMetadata)) {
+    // Flatten recusive LoggerWithMetadata instances
+    if (parent instanceof LoggerWithMetadata) {
+      this.parentLogger = parent.parentLogger;
+      this.metaChain = LoggerWithMetadata.appendToChain(parent.metaChain, meta);
+    } else {
+      this.parentLogger = parent;
+      this.metaChain = [meta];
+    }
+  }
+
+  log(level: LogLevel, message: string, meta?: LogMetadata): void {
+    this.parentLogger.log(level, message, this.resolveMetadata(meta));
+  }
+
+  trace(message: string, meta?: LogMetadata): void {
+    this.parentLogger.trace(message, this.resolveMetadata(meta));
+  }
+
+  debug(message: string, meta?: LogMetadata): void {
+    this.parentLogger.debug(message, this.resolveMetadata(meta));
+  }
+
+  info(message: string, meta?: LogMetadata): void {
+    this.parentLogger.info(message, this.resolveMetadata(meta));
+  }
+
+  warn(message: string, meta?: LogMetadata): void {
+    this.parentLogger.warn(message, this.resolveMetadata(meta));
+  }
+
+  error(message: string, meta?: LogMetadata): void {
+    this.parentLogger.error(message, this.resolveMetadata(meta));
+  }
+
+  private resolveMetadata(meta?: LogMetadata): LogMetadata {
```

Comment:
Why not do this eagerly instead of having to evaluate on every call?

Reply:
We do eagerly when that's possible (most of the time), but there are case where attributes on a sublogger may change after the sublogger's creation, in particular the case of workflow's and activity's context.